File: web/admin/src/api/tool.ts
import axios from 'axios';
import { APIResponse } from './types';

export interface Tool {
  name: string;
  description?: string;
  template_name: string;
  params: Record<string, string>;
}

const adminApiBase = '/api/admin';

// Define the API base URL
const toolApiBase = `${adminApiBase}/tools`;

// Create a Tool
export function createTool(tool: Tool): Promise<APIResponse<Tool>> {
  return axios
    .post<APIResponse<Tool>>(toolApiBase, tool)
    .then((res) => res.data);
}

// Get a Tool by name
export function getTool(name: string): Promise<APIResponse<Tool>> {
  return axios
    .get<APIResponse<Tool>>(`${toolApiBase}/${name}`)
    .then((res) => res.data);
}

// Update a Tool
export function updateTool(
  name: string,
  tool: Tool,
): Promise<APIResponse<Tool>> {
  return axios
    .put<APIResponse<Tool>>(`${toolApiBase}/${name}`, tool)
    .then((res) => res.data);
}

// Delete a Tool
export function deleteTool(name: string): Promise<APIResponse<void>> {
  return axios
    .delete<APIResponse<void>>(`${toolApiBase}/${name}`)
    .then((res) => res.data);
}

// List all Tools
export function listTools(): Promise<APIResponse<Tool[]>> {
  return axios.get<APIResponse<Tool[]>>(toolApiBase).then((res) => res.data);
}

File: internal/dao/tool.go
package dao

import "gorm.io/gorm"

// Tool (formerly CraftAtom) derived from craft template
type Tool struct {
	BaseModelWithoutPK
	Name         string            `json:"name" gorm:"primaryKey" binding:"required"`
	Description  string            `json:"description"`
	TemplateName string            `json:"template_name" binding:"required" `
	Params       map[string]string `json:"params" gorm:"serializer:json;index"`
}

func (Tool) TableName() string {
	return "tools"
}

// CreateTool creates a new Tool record
func CreateTool(db *gorm.DB, tool *Tool) error {
	return db.Create(tool).Error
}

// GetToolByName retrieves a Tool record by its Name
func GetToolByName(db *gorm.DB, name string) (*Tool, error) {
	var tool Tool
	result := db.Where("name = ?", name).First(&tool)
	return &tool, result.Error
}

// GetAllTools retrieves all Tool records
func GetAllTools(db *gorm.DB) ([]Tool, error) {
	var tools []Tool
	result := db.Find(&tools)
	return tools, result.Error
}

// UpdateTool updates an existing Tool record
func UpdateTool(db *gorm.DB, tool *Tool) error {
	return db.Save(tool).Error
}

// DeleteTool deletes a Tool record by its Name
func DeleteTool(db *gorm.DB, name string) error {
	tool, err := GetToolByName(db, name)
	if err != nil {
		return err
	}
	return db.Delete(tool).Error
}